marlowe-playground-client/src/Blockly/Dom.purs

```diff
@@ -0,0 +1,253 @@
+--             </block>
+--         </statement>
+--     </block>
+-- </xml>
+module Blockly.Dom where
+
+import Prelude
+import Blockly.Internal (workspaceToDom)
+import Blockly.Types (BlocklyState)
+import Control.Monad.Error.Extra (toMonadThrow)
+import Control.Monad.Except (runExceptT, throwError)
+import Control.Monad.Except.Trans (class MonadThrow)
+import Data.Array (find, length)
+import Data.Array.Partial as UnsafeArray
+import Data.Compactable (separate)
+import Data.Either (Either(..), note')
+import Data.Lens (Lens', _1, view)
+import Data.Lens.Iso.Newtype (_Newtype)
+import Data.Lens.Record (prop)
+import Data.Maybe (Maybe(..))
+import Data.Newtype (class Newtype)
+import Data.Symbol (SProxy(..))
+import Data.Traversable (traverse)
+import Data.Tuple (Tuple(..))
+import Effect (Effect)
+import Effect.Class (class MonadEffect, liftEffect)
+import Foreign.Object (Object)
+import Foreign.Object as Object
+import Partial.Unsafe (unsafePartial)
+import Web.DOM (Element, Node)
+import Web.DOM.Element as Element
+import Web.DOM.HTMLCollection as HTMLCollection
+import Web.DOM.Node as Node
+import Web.DOM.NodeList as NodeList
+import Web.DOM.ParentNode as ParentNode
+
+newtype Block
+  = Block
+  { id :: String
+  , type :: String
+  -- In the XML the children of a block are stored/represented as an array of elements, but to simplify
+  -- consumption we use a JS native Object (like a `Map String a` but with better performance).
+  -- This decision implies that we cannot have two childs properties with the same `name`, but I think we shouldn't
+  -- anyway, and if we do, we are going to have the same kind of error later on, while transforming from Dom -> Term
+  , children :: Object BlockChild
+  }
+
+derive instance newtypeBlock :: Newtype Block _
+
+_id :: Lens' Block String
+_id = _Newtype <<< prop (SProxy :: SProxy "id")
+
+data BlockChild
+  -- A Field is visually represented as a label and an editable field, for example
+  -- the "timeout" inside a "When block", or the value inside a "Constant block".
+  = Field String
+  -- A Value is visually represented as a label in the block that can be attached to a new block on the side.
+  -- For example "by", "the amount of" and "currency" in the "Deposit block"
+  | Value Block
+  -- A Statement is visually represented as the pluggable elements inside of a block. For example the contract inside
+  -- the root block or the Actions inside of a When block.
+  -- We represent the statement children as an Array of Blocks, but as you can see in the top level XML example,
+  -- each `<statement>` has a single block child, and siblings are represented by a `<next>` node nested inside.
+  -- In the initial version of this module, Next was a BlockChild constructor, making it easier to parse the
+  -- XML, but knowing that it would complicate the translation to Term, I decided to flatten the statement's childs
+  -- while parsing.
+  | Statement (Array Block)
+
+type ChildWithName
```

Comment:
could this be `NamedChild`? I feel like `NamedX` is quite common in programming and `XWithName` isn't